rfcs/text/0000-create-file-mach-o.md

```diff
@@ -0,0 +1,150 @@
+# 0000: Create the Mach-O sub-field of the File fieldset
+
+- Stage: **0 (strawperson)**
+- Date: **TBD**
+
+Create the Mach Object (Mach-O) sub-field, of the `file` top-level fieldset. This document metadata can be used for malware research, as well as coding and other application development efforts.
+
+## Fields
+
+**Stage 0**
+
+This RFC is to create the Mach-O sub-field within the `file.` fieldset. This will include 35 sub-fields.
+
+|   Name                                     |   Type     |   Description                                                               |
+|--------------------------------------------|------------|-----------------------------------------------------------------------------|
+|   file.mach-o.cpu                          |   object   |   CPU information for the file.                                             |
+|   file.mach-o.cpu.architecture             |   keyword  |   CPU architecture target for the file.                                     |
+|   file.mach-o.cpu.byte_order               |   keyword  |   CPU byte order for the file.                                              |
+|   file.mach-o.cpu.subtype                  |   keyword  |   CPU subtype for the file.                                                 |
+|   file.mach-o.cpu.type                     |   keyword  |   CPU type for the file.                                                    |
+|   file.mach-o.headers                      |   object   |   Header information for the file.                                          |
```

Comment:
Is the pluralization of this to account for fat archives? If so I'm assuming you want to actually move that up to the top level -- you generally have things like different cpu architectures that are associated with fat files. Instead, I'd just make this `header` or un-nest everything from under here and each `mach-o` entry be for a separate architecture. I'm assuming we might need to punt on fat architectures on the first pass (@webmat / @ebeahan  : not sure if it's a good idea to make the entire `mach-o` field set `nested`? any insight?).

Reply:
Is `nested` considered because these objects would need to be queried independently of each other? Nested (along with `flattened`) has been discussed before for various use cases, but ECS hasn't had a field adopt it yet.

How many possible architectures can be associated in this case? A small enough number that naming and nesting them into predictable places could be a consideration?

Reply:
So, @ebeahan , Apple's multi-architecture fat formats are a bit gross. I'm not entirely sure if macho can be a generic container for things like various MIPS architectures, but off the top of my head if we wanted to support what Mac architectures have been around in the past ~10 years (of which I'm sure there's a good amount of floating around out there), we'd likely need to have support for x86 32-bit and 64-bit architectures, ppc 32-bit and 64-bit architectures, and ARM 32-bit and 64-bit architectures.

Due to the ability to compile "fat" binaries that support multiple architectures, all of these could (theoretically) be in a single macho binary. Each would require all of the segments, symbols, headers fields at minimum since they're more or less independent of each other.

In practice, you're probably likely to run into fat binaries with 32-bit and 64-bit x86 and, with some older binaries, ppc64 architectures on a somewhat regular basis. With Apple's move to ARM chips now though, if they push continuing fat file formats, this could shift towards something like 64-bit x86 and ARM.

If we want to punt on the whole multi-architecture question we could possibly just say "forget the nested documents for now" and say that if you're parsing an object file with multiple formats in it, you should just emit 2 events, but IMO that's not exactly ideal.

Reply:
> Is the pluralization of this to account for fat archives? If so I'm assuming you want to actually move that up to the top level -- you generally have things like different cpu architectures that are associated with fat files. Instead, I'd just make this `header` or un-nest everything from under here and each `mach-o` entry be for a separate architecture. I'm assuming we might need to punt on fat architectures on the first pass.

Nothing that fancy, just an oversight. Fix't.

WRT to the "fat" binaries, I'm willing to put in the work if someone wants to guide me (or push into the PR themselves). if it's easier to pass for now, that's fine too.

Reply:
I've made the change to `header` from `headers`.

Was there a decision on nesting/un-nesting?

Reply:
Before making that call, I think there's a decision needed if these initial `mach-o` fields will capture header data for universal/fat macho-o binaries?

Once we understand and agree on the structure of possible events we need to model for, I think it'll be easier to agree on data types.

Reply:
So, it's probably important to consider multiple headers, unfortunately. It is gross because ideally we could have a similar structure across PE, Mach-O, and ELF...there are some fat binary adaptations for the other 2 platforms, but they're not strictly supported in the same way that the Mach-O format supports them.

Namely, I can imagine a visualization/aggregation where I can organize all my `file` document types by their exectuable header information...  i.e. executable type, cpu architecture, binary flags, etc. That said, the overlap of header specific information probably isn't insanely large. Looking at the [LIEF documentation](https://lief.quarkslab.com//doc/latest/api/python/abstract.html#header) (probably the most complete representation I've found), there's just a few fields in the abstract header, which would probably be useful in keyword arrays rather than nested objects (or both).

Reply:
So, based on this discussion and [this article](https://medium.com/@cyrilcermak/exploring-ios-es-mach-o-executable-structure-aa5d8d1c7103#f982) on fat binaries, it looks like they're going to be more prevalent in the future. It sounds like we agree moving to `headers` from `header` is the right approach.

@andrewstucki can you explain this? Do you mean to move this to `file.headers` as an object and things like `ARM` and `ARM64` being value examples?
> you want to actually move that up to the top level -- you generally have things like different cpu architectures that are associated with fat files

@dcode you mentioned that nested or keyword arrays would be helpful...do you have an opinion of which would be _more_ useful?

Reply:
So...  after digging deeper, it's worse than that. A fat Mach-o binary is literally two full binaries plus an extra header. That means, there are duplicate headers, sections, imports, exports, etc....  so...  long and short of it, not sure yet

Reply:
Looks like we're good to go with `macho.headers` as a nested field type. I've made the change to the docs.